keps/sig-windows/20190103-windows-node-support.md

```diff
@@ -173,6 +179,19 @@ Windows nodes will adhere to Kubernetes version-skew policy (node to control pla
 - [cncf-k8s-conformance thread](https://lists.cncf.io/g/cncf-k8s-conformance/topic/windows_conformance_tests/27913232)
 - [kubernetes/enhancements proposal](https://github.com/kubernetes/features/issues/116)
 
+### API Changes Proposed
+
+#### Add V1.SecurityContext.WindowsOptions.runAsUsername
```

Comment:
do any of the [existing fields in PodSecurityContext](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L2942-L2995) apply to Windows?

I'm in favor of grouping attributes, but wasn't sure if the existing security context object was essentially "LinuxSecurityContext" in nature, which would point more toward a `podSpec.windowsSecurityContext` bucket (more like the CRI ContainerConfig struct, which contains peer linux and windows config fields).

Reply:
The SecurityContext that apply / don't apply are called out in the Windows KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20190103-windows-node-support.md#api-reference

There is a tracking issue on PodSecurityPolicy: https://github.com/kubernetes/kubernetes/issues/64801 . Without runAsUsername, there won't be any Windows specific fields that would need an implementation in PodSecurityContext. The feedback I had in that issue was that PodSecurityContext would be replaced and not graduate to stable. If that's the case, I'd rather focus on whatever the new effort is (OPA?)

Reply:
> The SecurityContext that apply / don't apply are called out in the Windows KEP

The reference is helpful, thanks. I didn't see runAsNonRoot, sysctls, and procMount called out. Do those fundamentally not apply to windows? If they don't apply, then it seems like there are no fields inside pod.securityContext or container.securityContext that apply to windows.

> There is a tracking issue on PodSecurityPolicy: [kubernetes/kubernetes#64801](https://github.com/kubernetes/kubernetes/issues/64801)

My question wasn't about PodSecurityPolicy, but about the PodSecurityContext struct (the one that defines the structure of pod.spec.securityContext). I was wondering if it made more sense to nest a windows-specific struct underneath that, or if all the fields inside pod.spec.securityContext were linux-specific today, to use a peer pod.spec.windowsSecurityContext struct instead

Reply:
That's correct. None of the fields in PodSecurityContext apply to Windows. I'll do a separate small PR to call out those additional fields you mentioned. It doesn't change the scope of what we're doing for Windows stable.

The proposed changes in https://github.com/kubernetes/kubernetes/pull/73609 added a new field `PodSecurityContext.WindowsOptions`. What do you think of that compared to a peer struct? Personally I think it will be easier to find the right field in the API if it's under PodSecurityContext instead of a peer struct.

Reply:
> What do you think of that compared to a peer struct? Personally I think it will be easier to find the right field in the API if it's under PodSecurityContext instead of a peer struct.

On reflection, I think you're probably right. a helpful approach is to ask what a v2 pod would look like if we ever made one, and how we'd best position the current data to be converted to a more sensible layout. It seems likely we'd end up with at least one security-related field honored by both linux and windows, which would point in this direction:

```yaml
securityContext:
  ... shared fields that apply to linux and windows

  linuxOptions:
    ... current linux-specific fields

  windowsOptions:
    ... windows-specific fields
```